src/binfileutils.js

```diff
@@ -10,62 +14,143 @@ export async function readBinFile(fileName, type, maxVersion, cacheSize, pageSiz
     }
 
+    fd.binVersion = binVersion;
+
     return {fd, sections};
+
+    async function readBinFileV1() {
+        const nSections = await fd.readULE32();
+
+        // Scan sections
+        for (let i = 0; i < nSections; i++) {
+            let ht = await fd.readULE32();
+            let hl = await fd.readULE64();
+            if (typeof sections[ht] == "undefined") sections[ht] = [];
+            sections[ht].push({
+                p: fd.pos,
+                size: hl
+            });
+            fd.pos += hl;
+        }
+    }
+
+    async function readBinFileV2() {
+        //Reserved sections table size
+        const nReservedSections = await fd.readULE32();
+
+        //Get sections
+        for (let i = 0; i < nReservedSections; i++) {
+            //Section id
+            let sectionId = await fd.readULE32();
+            //Section size
+            let sectionSize = await fd.readULE64();
+            //Offset
+            let sectionOffset = await fd.readULE64();
+            if (sectionId !== 0) {
+                if (sections[sectionId] === undefined) sections[sectionId] = [];
+                sections[sectionId].push({
+                    p: sectionOffset,
+                    size: sectionSize,
+                });
+            }
+        }
+    }
 }
 
 export async function createBinFile(fileName, type, version, nSections, cacheSize, pageSize) {
-
     const fd = await fastFile.createOverride(fileName, cacheSize, pageSize);
 
     const buff = new Uint8Array(4);
     for (let i=0; i<4; i++) buff[i] = type.charCodeAt(i);
-    await fd.write(buff, 0); // Magic "r1cs"
+    await fd.write(buff, 0); // 4 bytes text indicating the file's type
 
     await fd.writeULE32(version); // Version
-    await fd.writeULE32(nSections); // Number of Sections
+
+    let [fileVersion, binVersion] = decodeVersion(version);
+
+    if(BIN_FORMAT_1 === binVersion) {
+        await fd.writeULE32(nSections); // Number of Sections
+    } else {
+        let nReservedSections = Math.ceil(nSections / SECTIONS_TABLE_STEP) * SECTIONS_TABLE_STEP;
+        await fd.writeULE32(nReservedSections); // Number of reserved sections
+
+        fd.pSectionsTable = fd.pos;
+        for (let i = 0; i < nReservedSections; i++) {
+            await fd.writeULE32(0); // Section type
+            await fd.writeULE64(0); // Section size
+            await fd.writeULE64(0); // Absolute offset
+        }
+    }
+
+    fd.binVersion = binVersion;
 
     return fd;
 }
 
 export async function startWriteSection(fd, idSection) {
-    if (typeof fd.writingSection !== "undefined") throw new Error("Already writing a section");
-    await fd.writeULE32(idSection); // Header type
-    fd.writingSection = {
-        pSectionSize: fd.pos
-    };
-    await fd.writeULE64(0); // Temporally set to 0 length
+    if (typeof fd.writingSection !== "undefined") {
+        throw new Error("Already writing a section");
+    }
+
+    if(BIN_FORMAT_1 === fd.binVersion) {
```

Comment:
Ah, I see that you didn't want to change the parameters here, which only works if you attached `binVersion` to the `fd` 🤔

Reply:
It's a flag for the file descriptor indicating which binary format version is using